### PR TITLE
Fix some inconsistency with terminology

### PR DESCRIPTION
**Pull request by floatn**
#579

### CODE REVIEW

File: nx/lib/nx.ex [hunk 1]
```diff
@@ -1573,21 +1573,54 @@ defmodule Nx do
...
       1
 
-      iex> Nx.to_scalar(Nx.tensor([1.0, 2.0, 3.0]))
-      ** (ArgumentError) cannot convert tensor of shape {3} to scalar
+      iex> Nx.to_number(Nx.tensor([1.0, 2.0, 3.0]))
+      ** (ArgumentError) cannot convert tensor of shape {3} to number
 
   """
   @doc type: :conversion
+  @deprecated "Use Nx.to_number/1 instead"
```
Comment (josevalim) [inline nx/lib/nx.ex:1623]: We don't need to deprecate it, you can go ahead and remove it. :)

File: nx/lib/nx/defn/compiler.ex [hunk 2]
```diff
@@ -48,7 +48,7 @@ defmodule Nx.Defn.Compiler do
 
   # These operations do not have valid meaning for Nx.Defn.Expr
   @forbidden_ops [:backend_copy, :backend_deallocate, :backend_transfer] ++
-                   [:to_binary, :to_scalar, :to_flat_list, :to_heatmap, :to_batched_list] ++
+                   [:to_binary, :to_number, :to_scalar, :to_flat_list, :to_heatmap, :to_batched_list] ++
```
Comment (josevalim) [inline nx/lib/nx/defn/compiler.ex:51]:
```suggestion
                   [:to_binary, :to_number, :to_flat_list, :to_heatmap, :to_batched_list] ++
```

**Comment by josevalim** [comment]
Thank you! I have dropped two minor comments. Also please update exla/torchx accordingly, the amount of changes there should be minimal. :)

File: nx/lib/nx.ex (hunk 1 above)
Comment (floatn) [reply nx/lib/nx.ex:1623]: Axon will not be happy.

Comment (josevalim) [reply nx/lib/nx.ex:1623]: Ok, let's keep this for a week and then we can change it. :)

Comment (floatn) [reply nx/lib/nx.ex:1623]: :) Alright.

File: nx/lib/nx/defn/compiler.ex (hunk 2 above)
Comment (floatn) [reply nx/lib/nx/defn/compiler.ex:51]: In a week.

Comment (josevalim) [inline nx/lib/nx/defn/compiler.ex:51]: This is annoying the formatter and without this the code will fail to compile anyway, so we are fine.

```suggestion
                   [:to_binary, :to_number, :to_flat_list, :to_heatmap, :to_batched_list] ++
```

**Review by josevalim — APPROVED** [review]

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**